### PR TITLE
[meshcop] publish Border Router ID in MeshCoP service

### PR DESCRIPTION
**Pull request by wgtdkp**
A randomly generated 16-bytes BR ID is required for adding Thread network to mobile OS APIs.
This commit generates the ID with OpenThread NonCrypto random API and save the ID in settings.

Example MeshCoP service dump  which includes the `id` field:
```
OpenThread BorderRouter #B465._meshcop._udp.local. state changed: ServiceStateChange.Updated
  Host: wgtdkp.local.
  Addresses: ['fe80::21bf:d2a9:a692:4196%0', 'fe80::7a95:56b6:37ca:9a18%0', '192.168.98.1', '192.168.97.37', '192.168.97.33', '192.168.97.29', '192.168.97.25', '192.168.97.21', '192.168.97.17', '192.168.97.13', '192.168.97.9', '192.168.97.5', '192.168.97.1', '192.168.96.1', '172.23.57.167', '100.84.187.27', '2401:fa00:41:801:456f:3951:5312:e779', '2401:fa00:41:801:abf2:84d:c187:b7fc', '2401:fa00:41:801:b3e6:d67a:b7e5:b337', '2401:fa00:41:801:f6af:37f8:9114:ea78', '2401:fa00:41:801:d5cd:356:54b8:b1b7', '2401:fa00:41:801:d3b:4457:29eb:ef0e', '2401:fa00:41:801:65ea:d70e:7f04:73cf', '2401:fa00:41:801:3ae4:3b3b:dd97:b03', '2401:fa00:41:23:8de3:6871:abc3:c5d7', '2401:fa00:41:23:2d0b:f1b6:ddd0:9827', '2401:fa00:41:23:1ce2:6043:a34f:e022', '2401:fa00:41:23:e4ed:dc8f:14c1:5b00', '2401:fa00:41:23:a9a4:380d:b0e1:6bb', '2401:fa00:41:23:9799:3cd3:bca5:c5d5', '2401:fa00:41:23:12a0:c083:71fd:4034', '2401:fa00:41:23:e21a:db0:6544:1068']
  Port: 49154
  Weight: 0
  Priority: 0
  TXT:
    rv: 1
    id: d6f4da2e14622c38f33ce44dcdc5cbc4
    vn: OpenThread
    mn: BorderRouter
    nn: OpenThread-d9a0
    xp: acc214689bc40bdf
    tv: 1.3.0
    xa: 7a61b152af5bb465
    sb: 000001b1
    at: ffeeddccbbaa0000
    pt: 1a416e35
    sq: 67
    bb: f0bf
    dn: DefaultDomain
    omr: 40fd5e55a6f0cc0001
```

Depends on https://github.com/openthread/openthread/pull/8971

### CODE REVIEW

File: src/border_agent/border_agent.cpp [hunk 1]
```diff
@@ -516,4 +525,19 @@ std::string BorderAgent::GetAlternativeServiceInstanceName() const
     return ss.str();
 }
 
+#if OTBR_ENABLE_PUBLISH_MESHCOP_BR_ID
+void BorderAgent::LoadOrGenerateId(void) {
+    otError error;
+    uint16_t idLength = sizeof(mId);
+
+    error = otPlatSettingsGet(mNcp.GetInstance(), OT_SETTINGS_KEY_BR_ID, 0, mId, &idLength);
```
Comment (abtink) [inline src/border_agent/border_agent.cpp:533]: I suggest we provide APIs for getting/saving this in OT instead? 
We can add APIs under `otBrAgent`?

Comment (wgtdkp) [reply src/border_agent/border_agent.cpp:533]: SGTM. @bukepo @jwhui thoughts?

File: src/border_agent/border_agent.cpp [hunk 2]
```diff
@@ -369,6 +371,24 @@ void BorderAgent::PublishMeshCopService(void)
...
     otbrLogInfo("Publish meshcop service %s.%s.local.", mServiceInstanceName.c_str(), kBorderAgentServiceType);
 
+#if OTBR_ENABLE_PUBLISH_MESHCOP_BA_ID
+    {
+        otError error;
+        uint8_t id[OT_BORDER_AGENT_ID_LENGTH];
+        uint16_t idLength = OT_BORDER_AGENT_ID_LENGTH;
+
+        error = otBorderAgentGetId(instance, id, &idLength);
+        if (error == OT_ERROR_NONE)
```
Comment (bukepo) [inline src/border_agent/border_agent.cpp:381]: should we verify idLength is 16?

File: src/border_agent/border_agent.cpp (hunk 1 above)
Comment (bukepo) [reply src/border_agent/border_agent.cpp]: SGTM

File: src/border_agent/border_agent.cpp (hunk 2 above)
Comment (wgtdkp) [reply src/border_agent/border_agent.cpp:381]: done

File: etc/cmake/options.cmake [hunk 3]
```diff
@@ -127,3 +127,8 @@ option(OTBR_DNS_UPSTREAM_QUERY "Allow sending DNS queries to upstream" OFF)
 if (OTBR_DNS_UPSTREAM_QUERY)
     target_compile_definitions(otbr-config INTERFACE OTBR_ENABLE_DNS_UPSTREAM_QUERY=1)
 endif()
+
+option(OTBR_PUBLISH_MESHCOP_BA_ID "Publish the MeshCoP mDNS 'id' TXT entry, enable this feature only when 'id' is not set via dbus API" ON)
```
Comment (wgtdkp) [inline etc/cmake/options.cmake:131]: @jwhui The option is enabled by default.

**Review by bukepo — APPROVED** [review]
nit: how about providing a command line argument to switch on/off instead of a compile time feature flag? really want to reduce compile time flags on POSIX platform.

**Review by abtink — APPROVED** [review]
LGTM. Thanks. 👍

**Comment by jwhui** [comment]
> nit: how about providing a command line argument to switch on/off instead of a compile time feature flag? really want to reduce compile time flags on POSIX platform.

What about also providing a CLI that allows setting the `id` value?

**Review by jwhui — APPROVED** [review]

**Comment by gabekassel** [comment]
> > nit: how about providing a command line argument to switch on/off instead of a compile time feature flag? really want to reduce compile time flags on POSIX platform.
> 
> What about also providing a CLI that allows setting the `id` value?

@wgtdkp would you like me to file a feature request?

**Comment by wgtdkp** [comment]
> > > nit: how about providing a command line argument to switch on/off instead of a compile time feature flag? really want to reduce compile time flags on POSIX platform.
> > 
> > 
> > What about also providing a CLI that allows setting the `id` value?
> 
> @wgtdkp would you like me to file a feature request?

Hey @gabekassel yeah I am aware of this, let me handle this in maybe this week :)

**Comment by wgtdkp** [comment]
> > nit: how about providing a command line argument to switch on/off instead of a compile time feature flag? really want to reduce compile time flags on POSIX platform.
> 
> What about also providing a CLI that allows setting the `id` value?

Submitting https://github.com/openthread/openthread/pull/9049 to address this

**Pull request merged**